### ISSUE TITLE
A `BrowserView` that can be used to embed additional web content into `webview` window

### ISSUE THREAD

**Issue by SOVLOOKUP** [tauri-apps/tauri#2709]
**Describe the solution you'd like**
A `BrowserView` that can be used to embed additional web content into `WebviewWindow` like [electron](https://www.electronjs.org/docs/latest/api/browser-view)

**Comment by wusyong** [comment]
Native webviews don't have such tag. It's a non-standard attribute.
There are others alternative that could also do the job such as iframe and shadow DOM.
They are also standard on the other hand.

**Comment by ghost** [comment]
> Native webviews don't have such tag. It's a non-standard attribute. There are others alternative that could also do the job such as iframe and shadow DOM. They are also standard on the other hand.

Iframes aren't really isolated or support every website. Webviews on the other hand do.

**Comment by xuchaoqian** [comment]
+1 for this feature.
It's very useful to support Embed WebView. With this feature, to create apps which support safety multitasking(such as TABs, multithread but still isolated) can be very easy.

**Comment by wusyong** [comment]
[iframe](https://caniuse.com/?search=iframe) and [shadow DOM](https://caniuse.com/?search=shadow%20Dom) are supported by all modern browsers while `BrowserView` is an electron specific feature.
I strongly recommend you consider these standard approaches. This could make your app compatible to all kind of browser engines.

**Comment by ghost** [comment]
> [iframe](https://caniuse.com/?search=iframe) and [shadow DOM](https://caniuse.com/?search=shadow%20Dom) are supported by all modern browsers while `BrowserView` is an electron specific feature. I strongly recommend you consider these standard approaches. This could make your app compatible to all kind of browser engines.

As I said, iframe doesn't support most sites and shadow DOM is not isolated.

**Comment by FabianLars** [comment]
@dragonDScript Don't ignore the first part of the earlier message:
> Native webviews don't have such tag.

And maybe additionally that one too:
> while BrowserView is an electron specific feature

Well i can see that this sentence(s) might not be quite informative/understandable enough (i blame the language barrier(s) lol). But it's still the most important message here, iframe and shadow dom are more or less just given as a consolidation.

Anyway, there are 2 difficulties here:
1) `BrowserView` is not a native html tag and we can't really create new ones "native" enough for truly embedded webviews -> the custom html tags most browsers already supports is not enough here. Especially since those aren't enabled in webkit by default _iirc_...
2) it's pretty much impossible (or close enough to impossible) to embed a webview inside a webview, iirc Microsoft themselves mentioned that somewhere in the WebView2Feedback repo (obv that's windows only, but i expect webkit to be way worse here). What might be (more) possible are 2 webviews side by side in the same window but again, without the html tag.

**The most important thing here is to keep in mind that tauri is not electron.** Electron has control over the whole browser architecture and can change everything they want about/in it. Tauri basically just wraps around a "browser" (3 different at the same time btw) and has to interact/work with what's there. (that was badly phrased but i hope you understand what i mean)
So in short and simplified: Electron is a browser, tauri is a wrapper around one.

So in the end it's not like we don't want it, it's more like it's not really possible at least not in the way electron users want it or are used to. If anyone **knows** that it indeed is possible in a sane and **cross-platform** way, then i'd appreciate resources about that (or even better a PR 👼 )

Edit: I don't want to say that it's truly and absolutely 100% impossible, i definitely could be in the wrong here, but it would be a really big surprise to me ngl (mainly because i already invested a little bit of time in researching this).

**Comment by niutech** [comment]
As @dragonDScript said, most of third-party websites are using the `X-Frame-Options` header and cannot be embedded inside an `<iframe>` tag. I would like to create a tiled web browser in Tauri with multiple websites shown in panes side-by-side. If something like [BrowserView](https://www.electronjs.org/docs/latest/api/browser-view) was available, I could create a sub-window stuck to some region of my main window (e.g. 50% x 100%) and auto-resize it whenever the main window resizes. Please reopen this issue as it is a useful feature.

**Comment by JonasKruckenberg** [comment]
As previously stated [here](https://github.com/tauri-apps/tauri/issues/2975#issuecomment-1019945044) and [here](https://github.com/tauri-apps/tauri/issues/2709) building browsers is something we don't support and we don't plan on supporting anytime soon. It just requires exorbitant amounts of engineering time to built something like this and have it be secure. https://byte.how/posts/webview-androids-most-convenient-footgun/ is an interesting article about why embedded webviews/iframes in webviews are a dangerous idea to begin with.

I would also like to point out that there is a reason why electrons `WebView` is deprecated for years now. It should not be used in electron either.

**Comment by ZeronoFreya** [comment]
> As @dragonDScript said, most of third-party websites are using the header and cannot be embedded inside an tag. I would like to create a tiled web browser in Tauri with multiple websites shown in panes side-by-side. If something like [BrowserView](https://www.electronjs.org/docs/latest/api/browser-view) was available, I could create a sub-window stuck to some region of my main window (e.g. 50% x 100%) and auto-resize it whenever the main window resizes. Please reopen this issue as it is a useful feature.`X-Frame-Options``<iframe>`

This should be feasible, create a borderless window as a sub-window, the size and position can be adapted according to a div node in the main window (requires the official to do this). The problem that users may encounter is the drag effect and sub-window occlusion. If possible, the sub-window is best located under the main window. Set the div node to pointer-events: none; to penetrate the main window, but this is not necessary.

**Comment by ghost** [comment]
> > As @dragonDScript said, most of third-party websites are using the header and cannot be embedded inside an tag. I would like to create a tiled web browser in Tauri with multiple websites shown in panes side-by-side. If something like [BrowserView](https://www.electronjs.org/docs/latest/api/browser-view) was available, I could create a sub-window stuck to some region of my main window (e.g. 50% x 100%) and auto-resize it whenever the main window resizes. Please reopen this issue as it is a useful feature.`` X-Frame-Options``<iframe> ``
> 
> This should be feasible, create a borderless window as a sub-window, the size and position can be adapted according to a div node in the main window (requires the official to do this). The problem that users may encounter is the drag effect and sub-window occlusion. If possible, the sub-window is best located under the main window. Set the div node to pointer-events: none; to penetrate the main window, but this is not necessary.

Tauri is not meant to be a browser

**Comment by ZeronoFreya** [comment]
> > > As @dragonDScript said, most of third-party websites are using the header and cannot be embedded inside an tag. I would like to create a tiled web browser in Tauri with multiple websites shown in panes side-by-side. If something like [BrowserView](https://www.electronjs.org/docs/latest/api/browser-view) was available, I could create a sub-window stuck to some region of my main window (e.g. 50% x 100%) and auto-resize it whenever the main window resizes. Please reopen this issue as it is a useful feature.` X-Frame-Options`<iframe> ``
> > 
> > 
> > This should be feasible, create a borderless window as a sub-window, the size and position can be adapted according to a div node in the main window (requires the official to do this). The problem that users may encounter is the drag effect and sub-window occlusion. If possible, the sub-window is best located under the main window. Set the div node to pointer-events: none; to penetrate the main window, but this is not necessary.
> 
> Tauri is not meant to be a browser

Nobody expects it to be a browser

**Comment by icoco** [comment]
if possible multiple web page in one Window  ?  we do not want replace exist chrome , we just want our app (base Tauri) provide web browser feature then user do not need open others

**Comment by valerius21** [comment]
I would find this feature to be quite useful.

**Comment by yenche123** [comment]
> I would find this feature to be quite useful.

@valerius21 
We all want it

**Comment by JiuRanYa** [comment]
This was quite useful for me

**Comment by seanoliver** [comment]
+1 we have a strong use case for this. Trying to migrate off Electron and onto Tauri.

**Comment by FabianLars** [comment]
We know this is a highly requested feature but please stop with the +1 comments. They don't matter to maintainers (a thumbs up works just as well) but instead just annoys normal users that are subscribed to this issue.

Also, please stop ignoring my lengthy comment above that explains why this feature won't be implemented anytime soon (if at all) :sweat_smile:

**Comment by zhouyangtingwen** [comment]
This was quite useful for me

**Comment by vaecou** [comment]
I also hope to have

**Comment by srsholmes** [comment]
Thanks so much for implementing this feature. 

In the future, do you think we would be able to control the webviews height and widths in using the JS bridge, so we could create a UI something along the lines of this? 

<img width="1492" alt="image" src="https://github.com/tauri-apps/tauri/assets/3579905/13d1a33d-07af-43a2-9e40-ef7c153a1423">

https://nomcopter.github.io/react-mosaic/

**Comment by niutech** [comment]
@srsholmes This grid layout has already been implemented in JS in [Mosaic](https://github.com/mlajtos/mosaic) browser (not the original one), but in my opinion it should be implemented not in JS, but natively in Rust/C++ for better performance, how I am doing in my [Split Browser](https://github.com/niutech/splitbrowser) (using Qt Advanced Docking System). For Rust, there are [egui_dock](https://lib.rs/crates/egui_dock) or [egui_tiles](https://lib.rs/crates/egui_tiles) with [tauri-egui](https://github.com/tauri-apps/tauri-egui).

**Comment by FabianLars** [comment]
> In the future, do you think we would be able to control the webviews height and widths in using the JS bridge, so we could create a UI something along the lines of this?

This should already be possible with the current implementation.

**Comment by srsholmes** [comment]
Thanks for the info / links @niutech . I'll check those projects out, they look really interesting. 

@FabianLars Thanks for the heads up. I havn't tried to interact with the windows via JS yet. I assumed they could only be interacted with by Rust as the example shows they were to be created on the rust side. I'll give it a go now.

**Comment by createchina** [comment]
Why can Electron use <webview> but Tauri can't? I need to seamlessly embed third-party web pages in my project and control or obtain the page data inside by injecting JavaScript.

**Comment by FabianLars** [comment]
> Why can Electron use but Tauri can't?

How often do i need to explain this? 😭 Electron uses chromium (or a fork of it) directly and therefore can modify the engine however they want. We use webviews (think of something like a browser that you can embedd in your app) over which we only have very limited control.

> I need to seamlessly embed third-party web pages in my project and control or obtain the page data inside by injecting JavaScript.

This should be possible with the multiwebview feature as well, maybe not just yet, but at least in theory.

**Comment by Elendiar** [comment]
@FabianLars Is it possible at list have basic control for webview, like refresh, forward/backward, scale etc?
In electron at least i can create custom menu for this.
After migrating to tauri i encountered a problem with this.

Workers must open different internal websites, and with default system browser this is inconvenient because it is impossible to switch between tabs, and with app webviews there is a lack of control over the window

![Image](https://github.com/user-attachments/assets/0e98a4ca-7a87-4450-b80b-4bff7232faa0)

**Comment by xuchaoqian** [comment]
> Is it possible at list have basic control for webview, like refresh, forward/backward, scale etc? In electron at least i can create custom menu for this. After migrating to tauri i encountered a problem with this.

It is definitely possible, since I have implemented these functions on my own. Unfortunately, there are still quite a few bugs when in multiple view mode, such as focus events not firing, keyboard events not firing, and proxied webviews causing unresponsiveness, etc. So if you want to build an app like a browser, it will be blocked by these issues.

To be honest, the multiple view mode is a very important feature. It opens up the possibility to develop lots of interesting apps, such as browsers, terminals, IDEs, etc. However, the priority of this feature seems a bit low right now.

**Comment by Elendiar** [comment]
> since I have implemented these functions on my own

Implement in tauri? How? Iframe is not an option because of origin block, evaluating js script that add buttons works only once until the page reloads

**Comment by xuchaoqian** [comment]
> Implement in tauri? How? Iframe is not an option because of origin block, evaluating js script that add buttons works only once until the page reloads

It needs to enable the unstable feature, just like [this example](https://github.com/tauri-apps/tauri/tree/dev/examples/multiwebview).

The idea is to use a header webview to control the content webviews. The control buttons, such as refresh, forward/backward, minimize, maximize, close, etc., sit on the header webview, and the web pages sit on the content webviews (you can create as many as you need). You can load any site on the content webviews.

**Comment by FabianLars** [comment]
>  In electron at least i can create custom menu for this.

Do you have a link to an example and/or docs for this? Not too deep into electron myself so idk what you're talking about.

**Comment by Elendiar** [comment]
> >  In electron at least i can create custom menu for this.
> 
> Do you have a link to an example and/or docs for this? Not too deep into electron myself so idk what you're talking about.

Here my old project that use electron  [Menu.buildFromTemplate](https://www.electronjs.org/docs/latest/api/menu)
to add native menu for window.

https://gitfront.io/r/Elendiar/21SQq6UV13ja/production-taskbar-client/blob/src/main/utils/createBrowserWindowTemplate.js

https://gitfront.io/r/Elendiar/21SQq6UV13ja/production-taskbar-client/blob/src/main/ipc/windowsChannel.js

Btw don't be horrified by some awful code with autohotkey compiled binary for some winapi, because it was the only way for stable work in nodejs without side effects like input slowdown or memory leaks in ffi callbacks. 
Implementing this with rust (windows-rs) was much more convenient and reliable.

**Comment by FabianLars** [comment]
Ohhh so more or less a normal native window menu? Tauri supports those too though not sure if icons work on the top level as well or just in submenus.

I wish i'd have a guide to link but rn i only have the tray guide on hand https://v2.tauri.app/learn/system-tray/#add-a-menu It uses the same menu apis so you could probably pull that part out and apply it on a window instead via https://v2.tauri.app/reference/javascript/api/namespacemenu/#setasappmenu / https://v2.tauri.app/reference/javascript/api/namespacemenu/#setaswindowmenu (also possible in rust)

Either way, thanks for the links :)

**Comment by Elendiar** [comment]
@FabianLars Thanks, i tried with native menu, its works, but i think multiwebview feature is better for this. For touch monitor users, you can render a menu (ui) larger and more convenient than the native one, the only thing is that it is not clear how to load (and bundle) it from another index.html file, otherwise loading from the main application route is not instantaneous
